examples/Raytracer_Benchmark/Raytracer.cpp

```diff
@@ -89,137 +89,219 @@ void InitializeModel(vecgeom::VPlacedVolume const *world, RaytracerData_t &rtdat
+
+    if (nextvol)
+      Raytracer::ApplyRTmodel(ray, snext, rtdata, generation);
+    else {
+      // Rays exiting directly must get the color contribution from the background
+      auto bkg_contrib = rtdata.fBkgColor;
+      float x          = ray.intensity.load();
+      bkg_contrib *= x;
+      ray.fColor += bkg_contrib;
+      ray.fDone = true;
+    }
   }
 
   return ray.fColor;
 }
 
-void ApplyRTmodel(Ray_t &ray, double step, RaytracerData_t const &rtdata)
+void ApplyRTmodel(Ray_t &ray, double step, RaytracerData_t const &rtdata, int generation)
 {
-  int depth = ray.fNextState.GetLevel();
-  if (rtdata.fModel == kRTspecular) { // specular reflection
-    // Calculate normal at the hit point
-    bool valid = ray.fVolume != nullptr && depth >= rtdata.fVisDepth;
-    if (valid) {
-      vecgeom::Transformation3D m;
-      ray.fNextState.TopMatrix(m);
-      auto localpoint = m.Transform(ray.fPos);
-      vecgeom::Vector3D<double> norm, lnorm;
-      ray.fVolume->GetLogicalVolume()->GetUnplacedVolume()->Normal(localpoint, lnorm);
-      m.InverseTransformDirection(lnorm, norm);
-      vecgeom::Vector3D<double> refl = ray.Reflect(norm);
-      refl.Normalize();
-      double calf = -rtdata.fSourceDir.Dot(refl);
-      // if (calf < 0) calf = 0;
-      // calf                   = vecCore::math::Pow(calf, fShininess);
-      auto specular_color = rtdata.fBkgColor;
-      specular_color.MultiplyLightChannel(1. + 0.5 * calf);
-      auto object_color = rtdata.fObjColor;
-      object_color.MultiplyLightChannel(1. + 0.5 * calf);
-      ray.fColor = specular_color + object_color;
-      ray.fDone  = true;
-      // std::cout << "calf = " << calf << "red=" << (int)ray.fColor.fComp.red << " green=" <<
-      // (int)ray.fColor.fComp.green
-      //          << " blue=" << (int)ray.fColor.fComp.blue << " alpha=" << (int)ray.fColor.fComp.alpha << std::endl;
-    }
-  } else if (rtdata.fModel == kRTtransparent) { // everything transparent 100% except volumes at visible depth
-    bool valid = ray.fVolume != nullptr && depth == rtdata.fVisDepth;
-    if (valid) {
-      float transparency = 0.85;
-      auto object_color  = rtdata.fObjColor;
-      object_color *= (1 - transparency);
-      ray.fColor += object_color;
+
+  auto lastvol = (Ray_t::VPlacedVolumePtr_t)ray.fCrtState.Top();
+  auto nextvol = (Ray_t::VPlacedVolumePtr_t)ray.fNextState.Top();
+
+  // Get material structure for last and next volumes
+  auto medium_prop_last = (MyMediumProp *)lastvol->GetLogicalVolume()->GetBasketManagerPtr();
+  auto medium_prop_next = (MyMediumProp *)nextvol->GetLogicalVolume()->GetBasketManagerPtr();
+
+  bool exit_transparent        = medium_prop_last->material == kRTtransparent;
+  bool transparent_transparent = medium_prop_next->material == kRTtransparent &&
+                                 medium_prop_last->material == kRTtransparent &&
+                                 (medium_prop_next->refr_index * medium_prop_last->refr_index) > 1.;
```

Comment:
Maybe the condition should be more strict here, you don't want to do refraction between 2 media having the same index, even if it is >1